File: examples/rust/src/bin/text.rs
//! Text: size, weight, style, decoration, alignment, spacing and markup.
//!
//! One string repeated with one property changed at a time. A property that
//! does nothing is a line that looks like the one above it, which is what a
//! showcase is for: `text_decoration` and a centred or right `text_align` both
//! drew exactly that, and both draw now. The two rows that still repeat their
//! neighbour are `text_stroke` and `paint_order`, which the binding underneath
//! cannot express -- its text style carries a colour and no stroke width.

use meo_canvas::{
    Box as BoxNode, Element, FlexDirection, Root, Styled, Text, TextAlign,
    TextDecoration, hex_rgb, px,
};
use meo_canvas_examples::{FONT, FORMATS, draw_with_fonts};

/// The same words every line draws.
const WORDS: &str = "Hxgp quick 0123";

/// One line at the family the example registers.
fn line(text: &str) -> Element {
    Text::new(text)
        .font_family(FONT.0)
        .font_size(16.0)
        .color(hex_rgb(0x14_14_1e))
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let column = |children: Vec<Element>| {
        BoxNode::new()
            .width(px(184.0))
            .padding(px(4.0))
            .flex_direction(FlexDirection::Column)
            .gap(px(3.0))
            .background_color(hex_rgb(0xf6_f6_f8))
            .children(children)
    };

    let left = column(vec![
        line(WORDS).font_size(11.0),
        line(WORDS).font_size(22.0),
        line(WORDS).bold(),
        line(WORDS).italic(),
        line(WORDS).text_decoration(TextDecoration::Underline),
        line(WORDS).text_decoration(TextDecoration::LineThrough),
        line(WORDS).letter_spacing(px(3.0)),
        line(WORDS).word_spacing(px(12.0)),
    ]);

    let right = column(vec![
        line(WORDS).text_align(TextAlign::Center),
        line(WORDS).text_align(TextAlign::Right),
        line(WORDS).line_height(2.0),
        // Markup: the parser turns the tags into runs, so the bold word is one
        // segment and the coloured one another.
        line("plain <b>bold</b> <color=#dc2828>red</color>"),
        // Rich text built from runs rather than parsed, which is the other way
        // to reach the same shape.
        Text::rich([
            ("two ".to_owned(), meo_canvas::Style::new()),
            ("runs".to_owned(), meo_canvas::Style::new().bold()),
        ])
        .font_family(FONT.0)
        .font_size(16.0)
        .color(hex_rgb(0x14_14_1e)),
        // A paragraph clamped to one line, with what replaces the rest.
        line("this line is far too long to fit in the width it is given")
            .max_lines(1)
            .ellipsis("…"),
        // A shadow and a stroke, which are paint rather than layout.
        line(WORDS).text_shadow(vec![meo_canvas::scene::TextShadow {
            offset_x: 2.0,
            offset_y: 2.0,
            blur: 2.0,
            color: meo_canvas::scene::Color::rgba(20, 20, 40, 140),
        }]),
        line(WORDS)
            .font_size(20.0)
            .color(hex_rgb(0xff_ff_ff))
            .text_stroke(meo_canvas::scene::TextStroke {
                width: 1.0,
                color: hex_rgb(0x14_14_1e),
            }),
        // The same stroke painted over the fill rather than under it, which is
        // only legible against the line above.
        line(WORDS)
            .font_size(20.0)
            .color(hex_rgb(0xff_ff_ff))
            .text_stroke(meo_canvas::scene::TextStroke {
                width: 1.0,
                color: hex_rgb(0x14_14_1e),
            })
            .paint_order(meo_canvas::scene::PaintOrder::Stroke),
        // Where a short line sits in the space its height leaves.
        line(WORDS)
            .line_height(2.0)
            .vertical_align(meo_canvas::scene::VerticalAlign::Bottom),
    ]);

    let root = Root::new(400.0, 300.0)
        .background_color(hex_rgb(0xff_ff_ff))
        .padding(px(8.0))
        .gap(px(6.0))
        .children([left, right]);

    draw_with_fonts("text", root, FORMATS, &[FONT])
}

File: examples/bun/src/text.ts
/**
 * Text: size, weight, style, decoration, alignment, spacing and markup.
 *
 * One string repeated with one property changed at a time. A property that does
 * nothing is a line that looks like the one above it, which is what a showcase
 * is for: `textDecoration` and a centred or right `textAlign` both drew exactly
 * that, and both draw now. The two rows that still repeat their neighbour are
 * `textStroke` and `paintOrder`, which the binding underneath cannot express —
 * its text style carries a colour and no stroke width.
 */

import { Box, RichText, Root, Text, type SceneNode } from 'meo-canvas'

import { FORMATS, draw } from './write.js'

/** The family this example registers, and the file behind it. */
const FONT = {
  family: 'Showcase',
  paths: ['../../crates/meo-canvas-core/tests/assets/fonts/Oswald-VariableFont_wght.ttf'],
}

/** The same words every line draws. */
const WORDS = 'Hxgp quick 0123'

/** One line at the family the example registers. */
const line = (text: string, rest: Record<string, unknown> = {}): SceneNode => Text(text, { fontFamily: FONT.family, fontSize: 16, color: '#14141e', ...rest })

const column = (children: readonly SceneNode[]): SceneNode =>
  Box({ width: 184, padding: 4, flexDirection: 'column', gap: 3, backgroundColor: '#f6f6f8', children: [...children] })

const canvas = await Root({
  width: 400,
  height: 300,
  backgroundColor: '#ffffff',
  padding: 8,
  gap: 6,
  fonts: [FONT],
  children: [
    column([
      line(WORDS, { fontSize: 11 }),
      line(WORDS, { fontSize: 22 }),
      line(WORDS, { fontWeight: 'bold' }),
      line(WORDS, { fontStyle: 'italic' }),
      line(WORDS, { textDecoration: 'underline' }),
      line(WORDS, { textDecoration: 'line-through' }),
      line(WORDS, { letterSpacing: 3 }),
      line(WORDS, { wordSpacing: 12 }),
    ]),
    column([
      line(WORDS, { textAlign: 'center' }),
      line(WORDS, { textAlign: 'right' }),
      line(WORDS, { lineHeight: 2 }),
      // Markup: the parser turns the tags into runs, so the bold word is one
      // segment and the coloured one another.
      line('plain <b>bold</b> <color=#dc2828>red</color>'),
      // Rich text built from runs rather than parsed, which is the other way to
      // reach the same shape.
      RichText(
        [
          { text: 'two ', style: {} },
          { text: 'runs', style: { fontWeight: 'bold' } },
        ],
        {
          fontFamily: FONT.family,
          fontSize: 16,
          color: '#14141e',
        },
      ),
      // A paragraph clamped to one line, with what replaces the rest.
      line('this line is far too long to fit in the width it is given', { maxLines: 1, ellipsis: true }),
      // A shadow and a stroke, which are paint rather than layout.
      line(WORDS, { textShadow: { offsetX: 2, offsetY: 2, blur: 2, color: '#1414288c' } }),
      line(WORDS, { fontSize: 20, color: '#ffffff', textStroke: { width: 1, color: '#14141e' } }),
      // The same stroke painted over the fill rather than under it, which is
      // only legible against the line above.
      line(WORDS, {
        fontSize: 20,
        color: '#ffffff',
        textStroke: { width: 1, color: '#14141e' },
        paintOrder: 'stroke',
      }),
      // Where a short line sits in the space its height leaves.
      line(WORDS, { lineHeight: 2, verticalAlign: 'bottom' }),
    ]),
  ],
})

await draw('text', canvas, FORMATS)
